craft_providers/lxd/lxd_instance.py

```diff
@@ -0,0 +1,491 @@
+        :param env: Additional environment to set for process.
+        :param kwargs: Additional keyword arguments to pass.
+
+        :returns: Popen instance.
+        """
+        return self.lxc.exec(
+            instance_name=self.name,
+            command=_rootify_lxc_command(command=command, env=env),
+            project=self.project,
+            remote=self.remote,
+            runner=subprocess.Popen,
+            **kwargs,
+        )
+
+    def execute_run(
+        self,
+        command: List[str],
+        env: Optional[Dict[str, Optional[str]]] = None,
+        **kwargs,
+    ) -> subprocess.CompletedProcess:
+        """Execute a command using subprocess.run().
+
+        The process' environment will inherit the execution environment's
+        default environment (PATH, etc.), but can be additionally configured via
+        env parameter.
+
+        :param command: Command to execute.
+        :param env: Additional environment to set for process.
+        :param kwargs: Keyword args to pass to subprocess.run().
+
+        :returns: Completed process.
+
+        :raises subprocess.CalledProcessError: if command fails and check is
+            True.
+        """
+        return self.lxc.exec(
+            instance_name=self.name,
+            command=_rootify_lxc_command(command=command, env=env),
+            project=self.project,
+            remote=self.remote,
+            runner=subprocess.run,
+            **kwargs,
+        )
+
+    def exists(self) -> bool:
+        """Check if instance exists.
+
+        :returns: True if instance exists.
+
+        :raises LXDError: On unexpected error.
+        """
+        return self._get_state() is not None
+
+    def _get_disk_devices(self) -> Dict[str, Any]:
+        """Query instance and return dictionary of disk devices."""
+        devices = self.lxc.config_device_show(
+            instance_name=self.name, project=self.project, remote=self.remote
+        )
+
+        disks = {}
+        for name, config in devices.items():
+            if config.get("type") == "disk":
+                disks[name] = config
+
+                # Ensure the expected keys are in config.
+                if "path" not in config or "source" not in config:
+                    raise LXDError(
+                        brief=f"Failed to parse lxc device {name!r}.",
+                        details=f"* Device configuration: {devices!r}",
+                    )
+
+        return disks
+
+    def _get_state(self) -> Optional[Dict[str, Any]]:
+        """Get state configuration for instance.
+
+        :returns: State information parsed from lxc if instance exists, else
+            None.
+
+        :raises LXDError: On unexpected error.
+        """
+        instances = self.lxc.list(project=self.project, remote=self.remote)
+
+        # lxc returns a filter instances starting with instance name rather
+        # than the exact instance.  Find the exact match...
```

Comment:
I understand what you're doing but this comment is confusing, perhaps rephrase it?

Reply:
Will do!

Reply:
Fixed